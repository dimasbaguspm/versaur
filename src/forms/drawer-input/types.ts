import type { DrawerProps } from '@/overlays/drawer/types'
import type { TextInputProps } from '../text-input/types'
import type { ChangeEventHandler, InputHTMLAttributes, ReactNode } from 'react'

export interface DrawerInputContextValue {
  isOpen: boolean
  setIsOpen: (open: boolean) => void
  onChange?: ChangeEventHandler<HTMLInputElement>
  value: InputHTMLAttributes<HTMLInputElement>['value']
}

/**
 * DrawerInputProps extends TextInputProps, but input is always readOnly
 * @property {boolean} open - Drawer open state (controlled)
 * @property {(open: boolean) => void} onOpenChange - Callback for drawer open state change
 * @property {React.ReactNode} children - Drawer content
 */
export interface DrawerInputProps
  extends Omit<TextInputProps, 'readOnly' | 'children'>,
    Pick<
      DrawerProps,
      | 'size'
      | 'position'
      | 'disableEscapeKeyDown'
      | 'disableOverlayClickToClose'
    > {
  children: (ctx: DrawerInputContextValue) => ReactNode
}
