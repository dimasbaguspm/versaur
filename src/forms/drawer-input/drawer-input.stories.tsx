/**
 * DrawerInput stories
 * Demonstrates usage and variations of DrawerInput for form selection in a drawer overlay
 * @group Forms/DrawerInput
 */
import { useState } from 'react'
import { DrawerInput } from './drawer-input'
import { Button } from '@/primitive/button/button'
import type { Meta, StoryObj } from '@storybook/react'
import { Text } from '@/primitive'
import { TextInput } from '../text-input'

const meta: Meta<typeof DrawerInput> = {
  title: 'Forms/DrawerInput',
  component: DrawerInput,
}
export default meta

type Story = StoryObj<typeof DrawerInput>

function BasicStory() {
  const [value, setValue] = useState('')
  return (
    <DrawerInput
      label='Select value'
      value={value}
      onChange={e => setValue(e.target.value)}
    >
      {ctx => (
        <>
          <DrawerInput.Header>
            <Text fontSize='lg' fontWeight='bold'>
              Drawer Input Example
            </Text>
          </DrawerInput.Header>

          <DrawerInput.Body>
            <TextInput
              value={value}
              onChange={ctx.onChange}
              placeholder='Type something...'
            />
          </DrawerInput.Body>
          <DrawerInput.Footer>
            <Button onClick={() => ctx.setIsOpen(false)}>Confirm</Button>
          </DrawerInput.Footer>
        </>
      )}
    </DrawerInput>
  )
}

export const Basic: Story = {
  render: BasicStory,
}
