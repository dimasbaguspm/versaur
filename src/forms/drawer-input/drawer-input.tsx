import { useState, useMemo, forwardRef } from 'react'
import { TextInput } from '../text-input/text-input'
import { Drawer } from '@/overlays/drawer/drawer'
import type { DrawerInputProps, DrawerInputContextValue } from './types'

/**
 * DrawerInput is a readOnly TextInput that opens a Drawer when clicked.
 * Drawer content is provided by consumer as children.
 */
const DrawerInputRoot = forwardRef<HTMLInputElement, DrawerInputProps>(
  (
    {
      children,
      onChange,
      value,
      size,
      position,
      disableEscapeKeyDown,
      disableOverlayClickToClose,
      ...textInputProps
    },
    ref
  ) => {
    const [isOpen, setIsOpen] = useState(false)

    const contextValue = useMemo(
      () => ({ isOpen, setIsOpen, onChange, value }),
      [isOpen, setIsOpen, onChange, value]
    ) satisfies DrawerInputContextValue

    return (
      <>
        <TextInput
          {...textInputProps}
          onChange={onChange}
          value={value}
          ref={ref}
          readOnly
          tabIndex={0}
          onClick={() => setIsOpen(true)}
          onKeyDown={e => {
            if (e.key === 'Enter' || e.key === ' ') {
              e.preventDefault()
              setIsOpen(true)
            }
          }}
          aria-haspopup='dialog'
          aria-expanded={isOpen}
        />
        <Drawer
          isOpen={isOpen}
          onClose={() => setIsOpen(false)}
          size={size}
          position={position}
          disableEscapeKeyDown={disableEscapeKeyDown}
          disableOverlayClickToClose={disableOverlayClickToClose}
        >
          {isOpen && children(contextValue)}
        </Drawer>
      </>
    )
  }
)

export const DrawerInput = Object.assign(DrawerInputRoot, {
  Header: Drawer.Header,
  Body: Drawer.Body,
  Footer: Drawer.Footer,
})
