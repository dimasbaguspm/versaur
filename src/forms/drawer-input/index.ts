export { DrawerInput } from './drawer-input'
export type * from './types'
