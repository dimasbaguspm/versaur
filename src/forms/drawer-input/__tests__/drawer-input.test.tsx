import { describe, it, expect } from 'vitest'
import { render, screen, fireEvent } from '@testing-library/react'
import { composeStories } from '@storybook/react'
import * as stories from '../drawer-input.stories'

const { Basic } = composeStories(stories)

describe('DrawerInput', () => {
  it('renders correctly and matches snapshot', () => {
    const { asFragment } = render(<Basic />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('opens drawer on input click and closes on confirm', () => {
    render(<Basic />)
    const input = screen.getByLabelText('Select value')
    fireEvent.click(input)
    expect(screen.getByRole('dialog')).toBeInTheDocument()
    const confirmBtn = screen.getByRole('button', { name: /confirm/i })
    fireEvent.click(confirmBtn)
    expect(screen.queryByRole('dialog')).not.toBeInTheDocument()
  })

  it('calls onChange when typing in TextInput', () => {
    render(<Basic />)
    const input = screen.getByLabelText('Select value')
    fireEvent.click(input)
    const textInput = screen.getByPlaceholderText('Type something...')
    fireEvent.change(textInput, { target: { value: 'hello' } })
    expect(textInput).toHaveValue('hello')
  })
})
