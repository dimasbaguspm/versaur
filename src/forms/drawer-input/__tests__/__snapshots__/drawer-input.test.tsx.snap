// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DrawerInput > renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <label
      class="block text-sm font-medium text-foreground mb-2"
      for="«r0»"
    >
      Select value
    </label>
    <div
      class="relative w-full"
    >
      <input
        aria-expanded="false"
        aria-haspopup="dialog"
        aria-invalid="false"
        class="block w-full rounded-md border bg-white transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none disabled:bg-gray-50 border-primary/30 text-foreground focus:border-primary focus:ring-primary/20 pl-3 pr-3 h-9"
        id="«r0»"
        placeholder="Click to open modal"
        readonly=""
        tabindex="0"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;
